blog/_posts/2021-08-20-more-prefab-tools.md

```diff
@@ -0,0 +1,177 @@
+Having components be both modular and allowing derived classes to inherit properly is necessary to prevent code duplication but a hard problem.
+My solution will likely work in the short term, but I'm hoping whoever inherits this code will give it a good think too in case there's a better way.
+
+#### You know... there's CSS for that
+One of the projects I didn't get to this summer was adding something like a layout manager to the prefab project.
+However, when I tried out ideas for a model, it became clear that CSS already has excellent methods for laying out components using [flexboxes](https://css-tricks.com/snippets/css/a-guide-to-flexbox/) and [grids](https://css-tricks.com/snippets/css/complete-guide-grid/).
+Rather than creating a whole new prefab component, Empirical could simply provide some artfully crafted classes in the [default prefab styles](https://github.com/devosoft/Empirical/blob/master/include/emp/prefab/DefaultPrefabStyles.less) that users could place on their own divs.
+In a similar way, `@media` queries are a great way to make dramatic changes a page's main layout or hide/show components without the costs of altering component hierarchies programmatically.
+In the end I've been pleasantly surprised with how many tricky programing problems can be avoided via artful CSS.
+
+## Thoughts on the Future
+Several challenges exist on the horizon for Empirical's prefabricated web tools.
+First, upgrading past Bootstrap v4 to v5 seems to break styling for a few components.
+This is unfortunate since Bootstrap v5 levelizes button and input groups, making them easier to construct and adds new components such as [offcanvas](https://getbootstrap.com/docs/5.1/components/offcanvas/) which is a fantastic place to put items when there is limited space on the page (i.e. on a mobile device).
+Second, there are plans to build the next version of Avida with some of these prefabricated web tools.
+The web components are based on Bootstrap, which is developed with mobile-first responsive design philosophy, where as Avida has traditionally had desktop application look-and-feel.
+Since nowadays the majority of people expect a web page to function on their mobile device and we are already using Bootstrap, it would be valuable to consider what opportunities might be afforded by developing future tools with mobile viewport in mind.
+In addition, a modernized style would likely be appealing to younger generations which would greatly benefit the next generation of Avida-ED.
+
+## Acknowledgements
+This work is supported through Active LENS: Learning Evolution and the Nature of Science using Evolution in Action (NSF IUSE #1432563).
+Any opinions, findings, and conclusions or recommendations expressed in this material are those of the author(s) and do not necessarily reflect the views of the National Science Foundation.
+
```

Comment:
I don't know much about how Empirical works under the hood, it honestly seems like magic running C++ on the web. Your post was really informative and well written! I especially liked all the gifs and references throughout your post, I may try to replicate some of it! Good job.